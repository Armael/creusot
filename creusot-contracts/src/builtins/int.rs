use std::ops::*;

#[rustc_diagnostic_item = "creusot_int"]
pub struct Int;

impl PartialEq for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "eq_int"]
    fn eq(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "ne_int"]
    fn ne(&self, _: &Int) -> bool {
        panic!()
    }
}

impl PartialOrd for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "le_int"]
    fn le(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "ge_int"]
    fn ge(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "lt_int"]
    fn lt(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "gt_int"]
    fn gt(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    fn partial_cmp(&self, _: &Self) -> Option<std::cmp::Ordering> {
        panic!()
    }
}

impl From<i32> for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "i32_to_int"]
    fn from(_: i32) -> Self {
        panic!()
    }
}

impl From<u32> for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "u32_to_int"]
    fn from(_: u32) -> Self {
        panic!()
    }
}

impl From<usize> for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "usize_to_int"]
    fn from(_: usize) -> Self {
        panic!()
    }
}

impl Add<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "add_int"]
    fn add(self, _: Int) -> Self {
        panic!()
    }
}

impl Sub<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "sub_int"]
    fn sub(self, _: Int) -> Self {
        panic!()
    }
}

impl Mul<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "mul_int"]
    fn mul(self, _: Int) -> Self {
        panic!()
    }
}

impl Div<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "div_int"]
    fn div(self, _: Int) -> Self {
        panic!()
    }
}

impl Rem<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "rem_int"]
    fn rem(self, _: Int) -> Self {
        panic!()
    }
}

impl Neg for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "neg_int"]
    fn neg(self) -> Self {
        panic!()
    }
}
