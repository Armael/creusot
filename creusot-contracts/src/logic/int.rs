use super::model::*;
use crate as creusot_contracts;
use crate::logic::*;
use creusot_contracts_proc::*;

use std::ops::*;

#[rustc_diagnostic_item = "creusot_int"]
pub struct Int(*mut ());

impl WellFounded for Int {}

impl Int {
    #[trusted]
    #[logic]
    #[creusot::builtins = "int.Power.power"]
    pub fn pow(self, _: Int) -> Int {
        absurd
    }
}

impl From<u8> for Int {
    #[logic]
    #[trusted]
    #[creusot::builtins = "prelude.UInt8.to_int"]
    fn from(_: u8) -> Self {
        absurd
    }
}
impl Model for u8 {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "prelude.UInt8.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<u32> for Int {
    #[logic]
    #[trusted]
    #[creusot::builtins = "mach.int.UInt32.to_int"]
    fn from(_: u32) -> Self {
        absurd
    }
}
impl Model for u32 {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "mach.int.UInt32.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<u64> for Int {
    #[logic]
    #[trusted]
    #[creusot::builtins = "mach.int.UInt64.to_int"]
    fn from(_: u64) -> Self {
        absurd
    }
}
impl Model for u64 {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "mach.int.UInt64.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<i64> for Int {
    #[logic]
    #[trusted]
    #[creusot::builtins = "mach.int.Int64.to_int"]
    fn from(_: i64) -> Self {
        absurd
    }
}
impl Model for i64 {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "mach.int.Int64.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<i32> for Int {
    #[logic]
    #[trusted]
    #[creusot::builtins = "mach.int.Int32.to_int"]
    fn from(_: i32) -> Self {
        absurd
    }
}
impl Model for i32 {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "mach.int.Int32.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<usize> for Int {
    #[logic]
    #[trusted]
    #[creusot::builtins = "mach.int.UInt64.to_int"]
    fn from(_: usize) -> Self {
        absurd
    }
}
impl Model for usize {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "mach.int.UInt64.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<isize> for Int {
    #[logic]
    #[trusted]
    #[rustc_diagnostic_item = "isize_to_int"]
    fn from(_: isize) -> Self {
        absurd
    }
}
impl Model for isize {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "mach.int.Int64.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl Add<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[creusot::builtins = "add_int"]
    fn add(self, _: Int) -> Self {
        panic!()
    }
}

impl Sub<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[creusot::builtins = "sub_int"]
    fn sub(self, _: Int) -> Self {
        panic!()
    }
}

impl Mul<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[creusot::builtins = "mul_int"]
    fn mul(self, _: Int) -> Self {
        panic!()
    }
}

impl Div<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[creusot::builtins = "div_int"]
    fn div(self, _: Int) -> Self {
        panic!()
    }
}

impl Rem<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[creusot::builtins = "rem_int"]
    fn rem(self, _: Int) -> Self {
        panic!()
    }
}

impl Neg for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[creusot::builtins = "neg_int"]
    fn neg(self) -> Self {
        panic!()
    }
}
