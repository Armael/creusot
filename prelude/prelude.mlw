module CreusotPrelude
  type borrowed 'a = { current : 'a; final: 'a }

  let function ( *_ ) x = x.current
  let function ( ^_ ) x = x.final

  val borrow_mut (a : 'a) : borrowed 'a
    ensures { * result = a }
end
