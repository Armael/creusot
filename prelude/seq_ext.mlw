module SeqExt
use seq.Seq

function subsequence (s : seq 't) (i : int) (j : int) : seq 't = s[i..j]

function mk_empty () : seq 't = Seq.empty
end