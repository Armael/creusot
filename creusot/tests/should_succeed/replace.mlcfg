
module Core_Option_Option_Type
  type core_option_option_type 't =
    | Core_Option_Option_None_Type
    | Core_Option_Option_Some_Type 't
    
end
module Core_Ptr_NonNull_NonNull_Type
  use prelude.Opaque
  type core_ptr_nonnull_nonnull_type 't =
    | Core_Ptr_NonNull_NonNull_Type opaque_ptr
    
end
module Core_Marker_PhantomData_Type
  type core_marker_phantomdata_type 't =
    | Core_Marker_PhantomData_Type
    
end
module Core_Ptr_Unique_Unique_Type
  use Core_Marker_PhantomData_Type
  use Core_Ptr_NonNull_NonNull_Type
  type core_ptr_unique_unique_type 't =
    | Core_Ptr_Unique_Unique_Type (Core_Ptr_NonNull_NonNull_Type.core_ptr_nonnull_nonnull_type 't) (Core_Marker_PhantomData_Type.core_marker_phantomdata_type 't)
    
end
module Alloc_Boxed_Box_Type
  use Core_Ptr_Unique_Unique_Type
  type alloc_boxed_box_type 't 'a =
    | Alloc_Boxed_Box_Type (Core_Ptr_Unique_Unique_Type.core_ptr_unique_unique_type 't) 'a
    
end
module Alloc_Alloc_Global_Type
  type alloc_alloc_global_type  =
    | Alloc_Alloc_Global_Type
    
end
module Replace_Something_Type
  use mach.int.Int
  use mach.int.UInt32
  use Alloc_Alloc_Global_Type
  use Alloc_Boxed_Box_Type
  use Core_Option_Option_Type
  type replace_something_type  =
    | Replace_Something_Type uint32 (Core_Option_Option_Type.core_option_option_type (replace_something_type))
    
end
module Replace_Test_Interface
  use Replace_Something_Type
  val test [@cfg:stackify] (_a : Replace_Something_Type.replace_something_type) (b : Replace_Something_Type.replace_something_type) : ()
    
end
module Replace_Test
  use Replace_Something_Type
  let rec cfg test [@cfg:stackify] [#"../replace.rs" 8 0 8 44] (_a : Replace_Something_Type.replace_something_type) (b : Replace_Something_Type.replace_something_type) : ()
    
   = [@vc:do_not_keep_trace] [@vc:sp]
  var _0 : ();
  var _a_1 : Replace_Something_Type.replace_something_type;
  var b_2 : Replace_Something_Type.replace_something_type;
  var _3 : Replace_Something_Type.replace_something_type;
  {
    _a_1 <- _a;
    b_2 <- b;
    goto BB0
  }
  BB0 {
    _3 <- b_2;
    b_2 <- any Replace_Something_Type.replace_something_type;
    _a_1 <- _3;
    goto BB1
  }
  BB1 {
    goto BB2
  }
  BB2 {
    _0 <- ();
    goto BB3
  }
  BB3 {
    goto BB4
  }
  BB4 {
    return _0
  }
  
end
