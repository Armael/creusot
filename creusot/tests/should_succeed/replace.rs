struct Something {
	a: u32, b: Box<Something>
}

fn test (mut a: Something, b: Something) {
  a = b;

}

fn main () {}
