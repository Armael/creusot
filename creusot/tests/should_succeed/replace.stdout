module Replace
  use Ref
  use mach.int.Int
  use mach.int.Int32
  use mach.int.Int64
  use mach.int.UInt32
  use mach.int.UInt64
  use string.Char
  use floating_point.Single
  use floating_point.Double
  use prelude.Prelude

  scope Type
    type core_option_option 't = 
      | Core_Option_Option_None
      | Core_Option_Option_Some('t)
      
    predicate drop_core_option_option (drop_t : 't -> bool) (self : core_option_option 't) = 
      match (self) with
        | Core_Option_Option_None -> true
        | Core_Option_Option_Some(a) -> drop_t a
        end
    type something  = 
      | Something(uint32, core_option_option (something))
      
    predicate drop_something (self : something) = 
      let Something(a, b) = self in drop_uint a && drop_core_option_option drop_fix b
  end
  let rec cfg test (o_a_1 : Type.something) (o_b_2 : Type.something) : () = 
  var _0 : ();
  var a_1 : Type.something;
  var b_2 : Type.something;
  var _3 : Type.something;
  {
    a_1 <- o_a_1;
    b_2 <- o_b_2;
    goto BB0
  }
  BB0 {
    assume { Type.drop_something _3 };
    _3 <- b_2;
    assume { Type.drop_something a_1 };
    a_1 <- _3;
    goto BB1
  }
  BB1 {
    goto BB2
  }
  BB2 {
    assume { Type.drop_something _3 };
    _0 <- ();
    goto BB3
  }
  BB3 {
    assume { Type.drop_something b_2 };
    goto BB4
  }
  BB4 {
    assume { Type.drop_something a_1 };
    return _0
  }
  
  let rec cfg main () : () = 
  var _0 : ();
  {
    goto BB0
  }
  BB0 {
    _0 <- ();
    return _0
  }
  
end
